### PR TITLE
Improve syntax/logic coherence in all gatk4 plugins

### PR DESCRIPTION
**Pull request by maxulysse**
## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/gatk4/mergevcfs/main.nf [hunk 1]
```diff
@@ -22,13 +22,9 @@ process GATK4_MERGEVCFS {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def vcfs = vcf.collect{ "I=${it}"}.join(' ')
+    def reference_command = use_reference ? "D=${dict}" : ""
```
Comment (FriederikeHanssen) [inline modules/gatk4/mergevcfs/main.nf:26]:
```suggestion
    def reference_command = dict ? "D=${dict}" : ""
```

File: modules/gatk4/mergevcfs/main.nf [hunk 2]
```diff
@@ -8,9 +8,9 @@ process GATK4_MERGEVCFS {
         'quay.io/biocontainers/gatk4:4.2.5.0--hdfd78af_0' }"
 
     input:
-    tuple val(meta), path(vcfs)
-    path  ref_dict
-    val   use_ref_dict
+    tuple val(meta), path(vcf)
+    path  dict
+    val   use_reference
```
Comment (FriederikeHanssen) [inline modules/gatk4/mergevcfs/main.nf:13]:
```suggestion
```

File: modules/gatk4/baserecalibratorspark/main.nf [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    } else {
+        avail_mem = task.memory.giga
+    }
+    """
+    gatk --java-options "-Xmx${avail_mem}g" BaseRecalibratorSpark \\
+        -R $fasta \\
+        -I $input \\
+        $sites_command \\
+        $intervals_command \\
+        --tmp-dir . \\
```
Comment (FriederikeHanssen) [inline modules/gatk4/baserecalibratorspark/main.nf:43]: I am torn whether this should go in `modules.config` it's quite important, but technically not a file. Eitherway should be the same for all gatk modules

File: modules/gatk4/estimatelibrarycomplexity/main.nf [hunk 4]
```diff
@@ -32,12 +32,13 @@ process GATK4_ESTIMATELIBRARYCOMPLEXITY {
...
-        ${crams} \
-        -O ${prefix}.metrics \
-        --REFERENCE_SEQUENCE ${fasta} \
-        --VALIDATION_STRINGENCY SILENT \
-        --TMP_DIR . $args
+    gatk --java-options "-Xmx${avail_mem}g" EstimateLibraryComplexity \\
+        ${crams} \\
+        -O ${prefix}.metrics \\
+        --REFERENCE_SEQUENCE ${fasta} \\
+        --VALIDATION_STRINGENCY SILENT \\
```
Comment (FriederikeHanssen) [inline modules/gatk4/estimatelibrarycomplexity/main.nf:39]: same here, should be a `args` IMO (I think I added this myself 🙈 )

File: modules/gatk4/applyvqsr/main.nf [hunk 5]
```diff
@@ -8,23 +8,23 @@ process GATK4_APPLYVQSR {
...
+    path "versions.yml"              , emit: versions
 
     when:
     task.ext.when == null || task.ext.when
 
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    refCommand = fasta ? "-R ${fasta} " : ''
+    ref_command = fasta ? "-R ${fasta} " : ''
```
Comment (FriederikeHanssen) [inline modules/gatk4/applyvqsr/main.nf:27]: Propose: call a `fasta`, `fasta` everywhere or `reference`

**Comment by FriederikeHanssen** [comment]
while you are at it, the mixture of -I, INPUT, and --input is outright cruel for the eyes...

**Comment by maxulysse** [comment]
> while you are at it, the mixture of -I, INPUT, and --input is outright cruel for the eyes...

I was thinking at looking at all commands for all tools and get the same logic for all.
Do you think `-I` or `--INPUT`?

**Comment by FriederikeHanssen** [comment]
`--INPUT` ? More descriptive (also for the parameters that won't be so obvious)

File: modules/gatk4/calculatecontamination/main.nf [hunk 6]
```diff
@@ -9,7 +9,7 @@ process GATK4_CALCULATECONTAMINATION {
 
     input:
     tuple val(meta), path(pileup), path(matched)
-    val segmentout
+    val   segmentout
```
Comment (maxulysse) [inline modules/gatk4/calculatecontamination/main.nf:12]: @FriederikeHanssen I believe that this should go in modules.config in for $args

Comment (FriederikeHanssen) [reply modules/gatk4/calculatecontamination/main.nf:12]: yes I agree

Comment (maxulysse) [reply modules/gatk4/calculatecontamination/main.nf]: finally keeping it here as it generates extra output.

Comment (FriederikeHanssen) [reply modules/gatk4/calculatecontamination/main.nf]: ok, can you add a comment there for future me who will forget this

Comment (maxulysse) [reply modules/gatk4/calculatecontamination/main.nf]: no worries, I'm planning to add comments to explain more stuff around

File: tests/modules/gatk4/mutect2/main.nf [hunk 7]
```diff
@@ -28,6 +29,28 @@ workflow test_gatk4_mutect2_tumor_normal_pair {
     GATK4_MUTECT2_PAIR ( input, fasta, fai, dict, germline_resource, germline_resource_tbi, panel_of_normals, panel_of_normals_tbi )
 }
 
+workflow test_gatk4_mutect2_tumor_normal_pair_f1r2 {
+    input = [ [ id:'test', normal_id:'normal', tumor_id:'tumour' ], // meta map
```
Comment (FriederikeHanssen) [inline tests/modules/gatk4/mutect2/main.nf:33]: why does this work here? we also didn't set any id's in the test data as far as I know

File: modules/gatk4/applybqsrspark/main.nf [hunk 8]
```diff
@@ -0,0 +1,51 @@
+process GATK4_APPLYBQSR_SPARK {
+    tag "$meta.id"
+    label 'process_low'
+
+    conda (params.enable_conda ? "bioconda::gatk4=4.2.3.0" : null)
```
Comment (FriederikeHanssen) [inline modules/gatk4/applybqsrspark/main.nf:5]: _sad noises_ , but at least something works

File: modules/gatk4/baserecalibrator/main.nf [hunk 9]
```diff
@@ -9,41 +9,40 @@ process GATK4_BASERECALIBRATOR {
...
     tuple val(meta), path(input), path(input_index), path(intervals)
-    path fasta
-    path fai
-    path dict
-    path knownSites
-    path knownSites_tbi
+    path  fasta
+    path  fai
+    path  dict
+    path  known_sites
```
Comment (FriederikeHanssen) [inline modules/gatk4/baserecalibrator/main.nf:15]: meta.yml probably needs some naming updates

File: modules/gatk4/estimatelibrarycomplexity/main.nf [hunk 10]
```diff
@@ -9,21 +9,21 @@ process GATK4_ESTIMATELIBRARYCOMPLEXITY {
 
     input:
     tuple val(meta), path(cram)
```
Comment (FriederikeHanssen) [inline modules/gatk4/estimatelibrarycomplexity/main.nf:11]: to be fair, I think this tools runs on cram and bam, so also input, maybe?

File: modules/gatk4/filtermutectcalls/main.nf [hunk 11]
```diff
@@ -8,10 +8,10 @@ process GATK4_FILTERMUTECTCALLS {
         'quay.io/biocontainers/gatk4:4.2.5.0--hdfd78af_0' }"
 
     input:
-    tuple val(meta), path(vcf), path(tbi), path(stats), path(orientationbias), path(segmentation), path(contaminationfile), val(contaminationest)
-    path fasta
-    path fai
-    path dict
+    tuple val(meta), path(vcf), path(tbi), path(stats), path(orientationbias), path(segmentation), path(table), val(estimate)
```
Comment (FriederikeHanssen) [inline modules/gatk4/filtermutectcalls/main.nf:11]: also update to the emta.yml

File: modules/gatk4/gatherbqsrreports/main.nf [hunk 12]
```diff
@@ -8,7 +8,7 @@ process GATK4_GATHERBQSRREPORTS {
         'quay.io/biocontainers/gatk4:4.2.5.0--hdfd78af_0' }"
 
     input:
-    tuple val(meta), path(recal_table)
+    tuple val(meta), path(table)
```
Comment (FriederikeHanssen) [inline modules/gatk4/gatherbqsrreports/main.nf:11]: meta.yml

File: modules/gatk4/gatherpileupsummaries/main.nf [hunk 13]
```diff
@@ -31,11 +31,12 @@ process GATK4_GATHERPILEUPSUMMARIES {
...
     """
-    gatk --java-options "-Xmx${avail_mem}g" \
-        GatherPileupSummaries \
-        --sequence-dictionary ${dict} \
-        ${input} \
-        -O ${prefix}.pileupsummaries.table
+    gatk --java-options "-Xmx${avail_mem}g" GatherPileupSummaries \\
+        $input_list \\
+        --sequence-dictionary $dict \\
+        --O ${prefix}.pileupsummaries.table \\
```
Comment (FriederikeHanssen) [inline modules/gatk4/gatherpileupsummaries/main.nf:37]: not `-O` ?

File: modules/gatk4/genotypegvcfs/main.nf [hunk 14]
```diff
@@ -10,10 +10,10 @@ process GATK4_GENOTYPEGVCFS {
     input:
     tuple val(meta), path(gvcf), path(gvcf_index), path(intervals), path(intervals_index)
     path  fasta
-    path  fasta_index
-    path  fasta_dict
+    path  fai
+    path  dict
     path  dbsnp
-    path  dbsnp_index
+    path  dbsnp_tbi
```
Comment (FriederikeHanssen) [inline modules/gatk4/genotypegvcfs/main.nf:16]: meta.yml

File: modules/gatk4/haplotypecaller/main.nf [hunk 15]
```diff
@@ -26,23 +26,22 @@ process GATK4_HAPLOTYPECALLER {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def interval_option = intervals ? "-L ${intervals}" : ""
-    def dbsnp_option    = dbsnp ? "-D ${dbsnp}" : ""
-    def avail_mem       = 3
+    def interval_command = intervals ? "-L ${intervals}" : ""
+    def dbsnp_command = dbsnp ? "-D ${dbsnp}" : ""
```
Comment (FriederikeHanssen) [inline modules/gatk4/haplotypecaller/main.nf:30]: can we align this?

File: modules/gatk4/markduplicates/main.nf [hunk 16]
```diff
@@ -8,7 +8,7 @@ process GATK4_MARKDUPLICATES {
         'quay.io/biocontainers/gatk4:4.2.5.0--hdfd78af_0' }"
 
     input:
-    tuple val(meta), path(bams)
+    tuple val(meta), path(bam)
```
Comment (FriederikeHanssen) [inline modules/gatk4/markduplicates/main.nf:11]: meta.yml

File: modules/gatk4/markduplicatesspark/meta.yml [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+      description: Marked duplicates BAM file
+      pattern: "*.{bam}"
+  - metrics:
+      type: file
+      description: Duplicate metrics file generated by GATK
+      pattern: "*.{metrics.txt}"
+
+authors:
+  - "@ajodeh-juma"
+  - "@FriederikeHanssen"
```
Comment (FriederikeHanssen) [inline modules/gatk4/markduplicatesspark/meta.yml:63]: + "@maxulysee"

File: modules/gatk4/mutect2/main.nf [hunk 18]
```diff
@@ -33,28 +30,10 @@ process GATK4_MUTECT2 {
...
-        normals_command = ''
-
-    } else {
-        panels_command = " --germline-resource $germline_resource --panel-of-normals $panel_of_normals --f1r2-tar-gz ${prefix}.f1r2.tar.gz"
-        normals_command = '-normal ' + which_norm.join( ' -normal ')
-    }
+    def inputs = input.collect{ "--input ${it}"}.join(" ")
+    def interval_command = intervals ? "--intervals $intervals" : ""
+    def pon_command = panel_of_normals ? "--panel-of-normals $panel_of_normals" : ""
+    def gr_command = germline_resource ? "--germline-resource $germline_resource" : ""
```
Comment (FriederikeHanssen) [inline modules/gatk4/mutect2/main.nf:36]: <3

File: tests/modules/gatk4/applybqsrspark/test.yml [hunk 19]
```diff
@@ -0,0 +1,32 @@
+- name: gatk4 applybqsr test_gatk4_applybqsr_spark
+  command: nextflow run tests/modules/gatk4/applybqsrspark -entry test_gatk4_applybqsr_spark -c tests/config/nextflow.config -c ./tests/modules/gatk4/applybqsrspark/nextflow.config
+  tags:
+    - gatk4
+    - gatk4/applybqsrspark
+  files:
+    - path: output/gatk4/test.bam
+      md5sum: d088422be886dc8507ff97fcc7dd968a
+    - path: output/gatk4/versions.yml
+      md5sum: d5c6455d8a77aecc63f87c795fc3443e
```
Comment (FriederikeHanssen) [inline tests/modules/gatk4/applybqsrspark/test.yml:10]:
```suggestion
```

File: tests/modules/gatk4/applybqsrspark/test.yml [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+- name: gatk4 applybqsr test_gatk4_applybqsr_spark_intervals
+  command: nextflow run tests/modules/gatk4/applybqsrspark -entry test_gatk4_applybqsr_spark_intervals -c tests/config/nextflow.config -c ./tests/modules/gatk4/applybqsrspark/nextflow.config
+  tags:
+    - gatk4
+    - gatk4/applybqsrspark
+  files:
+    - path: output/gatk4/test.bam
+      md5sum: 4bfa18d651abd945e240b05e70107716
+    - path: output/gatk4/versions.yml
+      md5sum: cb4cb8a62e117b4adc643ae47883d53c
```
Comment (FriederikeHanssen) [inline tests/modules/gatk4/applybqsrspark/test.yml:21]:
```suggestion
```

File: tests/modules/gatk4/applybqsrspark/test.yml [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+- name: gatk4 applybqsr test_gatk4_applybqsr_spark_cram
+  command: nextflow run tests/modules/gatk4/applybqsrspark -entry test_gatk4_applybqsr_spark_cram -c tests/config/nextflow.config -c ./tests/modules/gatk4/applybqsrspark/nextflow.config
+  tags:
+    - gatk4
+    - gatk4/applybqsrspark
+  files:
+    - path: output/gatk4/test.cram
+      md5sum: 2e0bca197af4f043a4a85152e6edbe04
+    - path: output/gatk4/versions.yml
+      md5sum: 1efaa18be943bab4e4c54191d6eaa260
```
Comment (FriederikeHanssen) [inline tests/modules/gatk4/applybqsrspark/test.yml:32]:
```suggestion
```

File: tests/modules/gatk4/variantrecalibrator/main.nf [hunk 22]
```diff
@@ -2,73 +2,67 @@
...
+                    file(params.test_data['homo_sapiens']['genome']['res_1000g_omni2_5_hg38_21_vcf_gz'], checkIfExists: true),
+                    file(params.test_data['homo_sapiens']['genome']['res_1000g_phase1_snps_hg38_21_vcf_gz'], checkIfExists: true),
+                    file(params.test_data['homo_sapiens']['genome']['dbsnp_138_hg38_21_vcf_gz'], checkIfExists: true)
+                ], [
+                    file(params.test_data['homo_sapiens']['genome']['hapmap_3_3_hg38_21_vcf_gz_tbi'], checkIfExists: true),
+                    file(params.test_data['homo_sapiens']['genome']['res_1000g_omni2_5_hg38_21_vcf_gz_tbi'], checkIfExists: true),
+                    file(params.test_data['homo_sapiens']['genome']['res_1000g_phase1_snps_hg38_21_vcf_gz_tbi'], checkIfExists: true),
+                    file(params.test_data['homo_sapiens']['genome']['dbsnp_138_hg38_21_vcf_gz_tbi'], checkIfExists: true)
+                ], [
+                    'hapmap,known=false,training=true,truth=true,prior=15.0 hapmap_3.3.hg38.vcf.gz',
```
Comment (FriederikeHanssen) [inline tests/modules/gatk4/variantrecalibrator/main.nf:57]: I don't know why GATK made these design decisions, but I don't like it.

**Review by FriederikeHanssen — CHANGES_REQUESTED** [review]
🤯 ❤️ Massive effort. some minor things mainly meta.yml and some version md5sum. 🚀

File: tests/modules/gatk4/mutect2/main.nf (hunk 7 above)
Comment (maxulysse) [reply tests/modules/gatk4/mutect2/main.nf:33]: no idea, I'm assuming at some point the bam header contains this normal or tumour, but no idea how it arrived there...

File: modules/gatk4/gatherpileupsummaries/main.nf (hunk 13 above)
Comment (maxulysse) [reply modules/gatk4/gatherpileupsummaries/main.nf:37]: probably typo, let me check tests

**Comment by maxulysse** [comment]
I'm just giving up on `genomicsdbimport` and `genotypegvcfs` as I have no idea how they are working and how to use them, so if @GCJMackenzie or anyone else has any pointers for improvement, I'm all for it.

File: modules/gatk4/markduplicatesspark/main.nf [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+    def input_list = bam.collect{"--INPUT ${it}"}.join(' ')
+
+    def avail_mem = 3
+    if (!task.memory) {
+        log.info '[GATK MarkDuplicatesSpark] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
+    } else {
+        avail_mem = task.memory.giga
+    }
+    """
+    export SPARK_USER=spark3
```
Comment (heuermh) [inline modules/gatk4/markduplicatesspark/main.nf:35]: This won't work with conda, right?

Comment (FriederikeHanssen) [reply modules/gatk4/markduplicatesspark/main.nf:35]: probably not, the whole spark thing has been an absolute major headache. Also the newest GATK versions are not working but only 4.2.3 😞

Comment (heuermh) [reply modules/gatk4/markduplicatesspark/main.nf:35]: Well, let's be sure to put the blame on GATK, not on Spark itself 😉

**Comment by FriederikeHanssen** [comment]
New release just got dropped: https://github.com/broadinstitute/gatk/releases/tag/4.2.6.0

File: modules/gatk4/haplotypecaller/main.nf (hunk 15 above)
Comment (maxulysse) [reply modules/gatk4/haplotypecaller/main.nf:30]: I choose not to align, because I don't think we align prefix and args by default in template

Comment (FriederikeHanssen) [reply modules/gatk4/haplotypecaller/main.nf:30]: ok

File: tests/modules/gatk4/variantrecalibrator/main.nf (hunk 22 above)
Comment (maxulysse) [reply tests/modules/gatk4/variantrecalibrator/main.nf:57]: > New release just got dropped: https://github.com/broadinstitute/gatk/releases/tag/4.2.6.0

not on bioconda yet, so will be done in a separate one

**Comment by maxulysse** [comment]
> New release just got dropped: https://github.com/broadinstitute/gatk/releases/tag/4.2.6.0

not on bioconda yet, so will be done in a separate one

File: tests/modules/gatk4/variantrecalibrator/main.nf (hunk 22 above)
Comment (maxulysse) [reply tests/modules/gatk4/variantrecalibrator/main.nf:57]: same

**Comment by maxulysse** [comment]
> > > New release just got dropped: https://github.com/broadinstitute/gatk/releases/tag/4.2.6.0
> > 
> > 
> > not on bioconda yet, so will be done in a separate one
> 
> I see it here: https://anaconda.org/bioconda/gatk4

not here: https://bioconda.github.io/recipes/gatk4/README.html

**Comment by FriederikeHanssen** [comment]
> > > > New release just got dropped: https://github.com/broadinstitute/gatk/releases/tag/4.2.6.0
> > > 
> > > 
> > > not on bioconda yet, so will be done in a separate one
> > 
> > 
> > I see it here: https://anaconda.org/bioconda/gatk4
> 
> not here: https://bioconda.github.io/recipes/gatk4/README.html

🤨 next one then

File: tests/modules/gatk4/mutect2/main.nf (hunk 7 above)
Comment (maxulysse) [reply tests/modules/gatk4/mutect2/main.nf:33]: I guess we need to sort this part out in sarek, and maybe update the module here

File: modules/gatk4/createsomaticpanelofnormals/main.nf [hunk 24]
```diff
@@ -24,18 +24,18 @@ process GATK4_CREATESOMATICPANELOFNORMALS {
...
         avail_mem = task.memory.giga
     }
     """
-    gatk --java-options "-Xmx${avail_mem}g" \\
-        CreateSomaticPanelOfNormals \\
-        -R $fasta \\
-        -V gendb://$genomicsdb \\
-        -O ${prefix}.vcf.gz \\
+    gatk --java-options "-Xmx${avail_mem}g" CreateSomaticPanelOfNormals \\
+        --variant gendb://$genomicsdb \\
```
Comment (FriederikeHanssen) [inline modules/gatk4/createsomaticpanelofnormals/main.nf:36]: I guess a GATK thing to only use this DB?

**Review by FriederikeHanssen — APPROVED** [review]
Looks good. One more tiny thing I noticed now, not all modules have `--TMP-DIR/--tmp-dir`. Some tests still failing

File: modules/gatk4/createsomaticpanelofnormals/main.nf (hunk 24 above)
Comment (maxulysse) [reply modules/gatk4/createsomaticpanelofnormals/main.nf:36]: I'd assume so

**Comment by maxulysse** [comment]
> Looks good. One more tiny thing I noticed now, not all modules have `--TMP-DIR/--tmp-dir`. Some tests still failing

Good point, I'm on it

**Pull request merged**